bgls/utils.py

```diff
@@ -56,3 +61,157 @@ def cirq_density_matrix_bitstring_probability(
         density_matrix_state.target_tensor, num_qubits
     )
     return np.abs(density_matrix[index, index])
+
+
+def cirq_stabilizer_ch_bitstring_probability(
+    stabilizer_ch_form_state: cirq.sim.StabilizerChFormSimulationState,
+    bitstring: str,
+) -> float:
+    """Returns the probability of measuring the `bitstring` (|z⟩) in the
+    `cirq.StabilizerChFormSimulationState` (U_C U_H|s⟩), i.e. |⟨z|ψ⟩|^2.
+
+    Args:
+        stabilizer_ch_form_state: Stabilizer state in CH form (arxiv:
+        1808.00128) as a 'cirq.StabilizerChFormSimulationState'.
+        bitstring: Bitstring |z⟩ as a binary string.
+
+    """
+    # the state is of type StabilizerStateChForm
+    state = stabilizer_ch_form_state.state
+    index = int(bitstring, 2)
+    # this runs in O(n^2) for an n qubit state
+    return np.abs(state.inner_product_of_state_and_x(index)) ** 2
+
+
+def apply_near_clifford_gate(
+    op: cirq.Operation,
+    state: bgls.simulator.State,
+    rng: np.random.RandomState = np.random.RandomState(),
+) -> None:
+    """
+    Applies a Clifford+T gate to a state. If the gate is Clifford,
+    apply normally, else choose one of the gates in the Clifford-expansion
+    to apply.
+
+    Parameters:
+        op: operation to apply.
+        state: simulator state to update.
+        rng: rng for controlling gate-expansion decisions.
+    """
+    if cirq.has_stabilizer_effect(op):
+        cirq.protocols.act_on(op, state)
+    else:
+        # assuming T gate
```

Comment:
If you're assuming this, you should assert it and throw and error if it's not true.

Reply:
Got it, figured out how to ensure the gate is T and it works now